### ISSUE TITLE
[style] Revisit enum sections in the style guide and consider adding more guidance around casts

### ISSUE THREAD

**Issue by msfschaffner** [lowRISC/opentitan#9569]
As recent discussions (#9391) have shown, we may need more guidance around enum casting in our style guide.

This issue is intended for collecting concerns and ideas so that they can eventually be translated to recommendations.

1. One concern surfaced in #9391 is how tools will actually synthesize a cast from a logic vector to an enum type of the same size, but with fewer allowed states. I.e., will the synthesis tools "filter" any out of range value and collapse them to one enum representation. This is spun out into its own issue (#9568), since we need feedback from tooling vendors for this.

1. Another concern recently brought up by @tjaychen is (please add more detail if necessary): How do we catch enum to enum typecasts where the length is mismatched. A cast will silence associated lint warning, since the conversion is made explicit. At the same time, the explicit cast may cut off bits if the width is mismatched - and this can go unnoticed.

**Comment by rswarbrick** [comment]
Personally, I don't think we should be passing enum types around at all, so I'm not sure I'm qualified to participate in this style discussion :-D

**Comment by GregAC** [comment]
> Another concern recently brought up by @tjaychen is (please add more detail if necessary): How do we catch enum to enum typecasts where the length is mismatched. A cast will silence associated lint warning, since the conversion is made explicit. At the same time, the explicit cast may cut off bits if the width is mismatched - and this can go unnoticed.

Could we disallow enum to enum typecasts entirely? I'd imagine we're not using many, can @tjaychen point to any examples?

**Comment by msfschaffner** [comment]
I think most of our enum typecasts are from logic vectors to enums and vice versa. Enum to enum is rather an exception, I believe...

**Comment by tjaychen** [comment]
> I think most of our enum typecasts are from logic vectors to enums and vice versa. Enum to enum is rather an exception, I believe...

yeah agreed. I think one of the biggest places where this is used is actually out of regfile (since our reggen doesn't currently understand enums). Long term I think it would be nice to absorb that understanding in.

I guess the other place this is used is mostly for padding out logic vectors like @msfschaffner pointed out.
